File: hashtable/36. Valid Sudoku.py
from typing import List
import collections

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = collections.defaultdict(set)
        col = collections.defaultdict(set)
        squares = collections.defaultdict(set)

        for i in range(9):
            for j in range(9):

                c = board[i][j]
                if c == '.':  continue
                if c not in row[i]:
                    row[i].add(c)
                else:
                    return False
                
                if c not in col[j]:
                    col[j].add(c)
                else:
                    return False
                

                if c not in squares[i//3 + (j//3)*3]:
                    squares[i//3 + (j//3)*3].add(c)
                else:
                    return False
        
        return True

File: hashtable/36. Valid Sudoku.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int size = board.size();
        vector<unordered_map<char, int>> col(size), row(size);
        vector<vector<unordered_map<char, int>>> block(3, vector<unordered_map<char, int>>(3));

        for(int i = 0; i < size; i++) {
            for(int j = 0; j < size; j++) {

                char c = board[i][j];
                if(c == '.')  continue;
                else{
                    if(row[i][c]++ > 0 | col[j][c]++ > 0 | block[i/3][j/3][c]++ > 0)  return false;
                }
            }
        }
    }
};
/*
[".",".","4",".",".",".","6","3","."],
[".",".",".",".",".",".",".",".","."],
["5",".",".",".",".",".",".","9","."],
[".",".",".","5","6",".",".",".","."],
["4",".","3",".",".",".",".",".","1"],
[".",".",".","7",".",".",".",".","."],
[".",".",".","5",".",".",".",".","."],
[".",".",".",".",".",".",".",".","."],
[".",".",".",".",".",".",".",".","."]*/